swarm/src/connection.rs

```diff
@@ -479,6 +481,19 @@ fn gather_supported_protocols(handler: &impl ConnectionHandler) -> HashSet<Strea
         .collect()
 }
 
+/// Constructs a [`Delay`] for the given [`Duration`] and returns the [`Instant`] at which it will fire.
+///
+/// If [`Duration`] + [`Instant::now`] overflows, we will return a [`Delay`] that at least sleeps _very_ long.
+fn sleep_until_or_at_least_very_long(mut duration: Duration) -> (Delay, Instant) {
+    while Instant::now().checked_add(duration).is_none() {
+        log::debug!("Cannot represent time {duration:?} in the future, halving it ...");
+
+        duration /= 2;
+    }
+
+    (Delay::new(duration), Instant::now() + duration)
+}
```

Comment:
Quite a hack. That said, I can not think of an easier fix.

Given that this is happening in a `poll` loop, would it make sense to reduce the calls to `Instant::now()`? E.g. go from 3 calls to one at the start of the method? We can't get around the call within `Delay::new`.

Reply:
Yeah that makes sense! It is also cleaner to always compare against the same one.